### PR TITLE
#184347045 remove expired product

### PR DESCRIPTION
**Pull request by Aline-Uwera**
### What does this PR do?
The products should stop listing if the expiry date has passed

### Description of Task to be completed?
Given the expiration date has passed, the product should be removed from
the list of available products
### How should this be manually tested?
1. Checkout to this branch: feature/delist-expired-products-184347045
2. Run the command npm install to install packages
3. Run `npm run dev`
4. Check functionalities of cron job for changing availability of expired products
#### What are the relevant pivotal tracker stories?
[184347045](https://www.pivotaltracker.com/story/show/184347045)

### CODE REVIEW

**Review by irakozetony — APPROVED** [review]

File: src/controllers/product.controller.js [hunk 1]
```diff
@@ -21,3 +24,24 @@ export const createProduct = async (req, res) => {
       });
   }
 };
+export const removeExpiredProducts = async (req, res) => {
+  try {
+    const isExpired = req.body.isExpired;
+    const isAvailable = req.body.isAvailable;
+    const productId = req.params.productId;
```
Comment (TuyizeeAnastase) [inline src/controllers/product.controller.js]: use destructuring assignment

File: src/controllers/product.controller.js [hunk 2]
```diff
@@ -21,3 +24,24 @@ export const createProduct = async (req, res) => {
...
+    await ProductService.updateProduct({ isExpired, isAvailable },productId );
+    return res.status(200).json({message:'Expired product was removed successfully'});
+  } catch (error) {
+    return res
+      .status(500)
+      .json({
+        error: error.message,
+        message: 'Failed to remove product from list',
```
Comment (TuyizeeAnastase) [inline src/controllers/product.controller.js]: message:"Failed to remove expired product from list

File: src/routes/api/product.routes.js [hunk 3]
```diff
@@ -23,4 +23,12 @@ productRoutes.post(
   uploadImages,
   createProduct
 );
+
+productRoutes.patch(
+  '/:productId',
+  protectRoute,
+  restrictTo('seller','admin'),
+  removeExpiredProducts,
+);
```
Comment (TuyizeeAnastase) [inline src/routes/api/product.routes.js]: - Here you need middleware to check if the product exists
- and middleware to check if the seller is the owner of the product

File: src/services/delistExpiredProduct.service.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import { ProductService } from './product.service';
+import { UserService } from './user.service';
+
+
+export const delistExpiredProducts = async () => {
+    const products = await models.Products.findAll();
+    for (const product of products) {
+        
+        if (product.expiryDate < new Date()) {
+            const isExpired =product.isExpired = true;
```
Comment (TuyizeeAnastase) [inline src/services/delistExpiredProduct.service.js]: what is the use of this line of code

**Review by TuyizeeAnastase — CHANGES_REQUESTED** [review]

File: src/routes/api/product.routes.js [hunk 5]
```diff
@@ -28,4 +31,13 @@ productRoutes.post(
 );
 
 productRoutes.get('/', protectRoute, getAllProducts);
+productRoutes.patch(
+  '/:productId',
```
Comment (ncutixavier) [inline src/routes/api/product.routes.js]: Change route to `/:productId/expired`

File: src/middlewares/product.middleware.js [hunk 6]
```diff
@@ -41,3 +43,29 @@ export const checkNumberOfImages = (req, res, next) => {
...
+
+  const authToken =  req.header('Authorization') || '';
+  const userToken = authToken.split(' ')[1];
+  const loggedUser = JwtUtility.verifyToken(userToken);
+
+  if (loggedUser.id !== sellerId){
+    res.status(400).json({ message: 'Product is owned by other seller' });
+  }else {
+    next();
+  }
```
Comment (ncutixavier) [inline src/middlewares/product.middleware.js]: Use `protectRoute` to get logged in user

File: src/services/delistExpiredProduct.service.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+import { ProductService } from './product.service';
+import { UserService } from './user.service';
+
+
+export const delistExpiredProducts = async () => {
+    const products = await models.Products.findAll();
+    for (const product of products) {
+        
+        if (product.expiryDate < new Date()) {
+            const isExpired = product.isExpired = true;
```
Comment (ncutixavier) [inline src/services/delistExpiredProduct.service.js]: Refactor this line

File: src/utils/mailTemplates.util.js [hunk 8]
```diff
@@ -26,3 +26,7 @@ export const login2FATemplate = (otpCode, firstname, id)=>{
   </div>
 </body>`;
 };
+
+export const expiredProductMessage= (product) =>{
+ return `The product " ${product.name}" has expired. Please remove it from the list of available products.`;
+};
```
Comment (ncutixavier) [inline src/utils/mailTemplates.util.js]: Update this message

**Review by ncutixavier — CHANGES_REQUESTED** [review]

**Pull request closed without merging**